src/wallet/home_cards/deferred_deep_link/parser.rs

```diff
@@ -0,0 +1,286 @@
+use crate::prelude::*;
+use base64::engine::general_purpose::URL_SAFE_NO_PAD;
+use base64::Engine;
+
+#[async_trait::async_trait]
+pub trait DeferredDeepLinkParser: Send + Sync {
+    async fn parse(&self, encoded_value: String) -> Result<HomeCards>;
+}
+
+pub struct Parser {
+    gateway_client: GatewayClient,
+}
+
+impl Parser {
+    pub fn new(gateway_client: GatewayClient) -> Self {
+        Self { gateway_client }
+    }
+}
+
+#[async_trait::async_trait]
+impl DeferredDeepLinkParser for Parser {
+    async fn parse(&self, encoded_value: String) -> Result<HomeCards> {
+        let decoded = self.decode(encoded_value)?;
+        self.transform_onboarding_deep_link_value(decoded).await
+    }
+}
+
+impl Parser {
+    fn decode(
+        &self,
+        encoded_value: impl AsRef<str>,
+    ) -> Result<OnboardingDeepLinkValue> {
+        let decoded_value_json_bytes = URL_SAFE_NO_PAD
+            .decode(encoded_value.as_ref())
+            .map_err(|e| {
+                println!("{}", e);
+                CommonError::DeferredDeepLinkInvalidValueFormat
+            })?;
+
+        serde_json::from_slice::<OnboardingDeepLinkValue>(
+            decoded_value_json_bytes.as_ref(),
+        )
+        .map_err(|_| CommonError::DeferredDeepLinkDecodingFailed)
+    }
+}
+
+#[cfg(test)]
```

Comment:
Better for tests to be at the bottom, I was confused finding non test code after test code :)

Reply:
Yeah, I can see why this is confusing, will move. My idea was following some idea that was mentioned on the course about placing the tests next to the function/feature tested. However, it definitely doesn't help when looking at a single file.

Unrelated, but one thing that bothers me a bit about our project is that every time I want to search for how something similar is done/used, I mostly get test search results. Then you need to manually filter among each results for the cases that are actually production code.
This could be resolved if tests were placed under a separate file with `_tests.rs` suffix. This way, you could easily filter them out on searches.

Reply:
about tests placement - we will likely have separate files for tests once we will move to have a workspace with multiple crates, then each crate will have at top level two folders `src` and `tests`.